ra2ce/graph/osm_network_wrapper.py

```diff
@@ -0,0 +1,266 @@
+            extremities_data = ExtremitiesData.get_extremities_data_for_sub_graph(from_node_id=u, to_node_id=v,
+                                                                                  sub_graph=unique_graph,
+                                                                                  graph=graph,
+                                                                                  shared_elements=unique_elements)
+            if extremities_data.from_id != extremities_data.to_id:  # excluding a self-directed link
+                if extremities_data.from_to_id is not None and extremities_data.from_to_coor is not None:
+                    if (extremities_data.from_to_id, extremities_data.to_from_id) not in unique_elements and \
+                            (extremities_data.from_to_coor, extremities_data.to_from_coor) not in unique_elements:
+                        edge_attributes = {
+                            key: value for key, value in data.items()
+                        }
+                        unique_graph.add_edge(extremities_data.from_id, extremities_data.to_id, **edge_attributes)
+                        unique_elements.add((extremities_data.from_to_id, extremities_data.to_from_id))
+                        unique_elements.add((extremities_data.from_to_coor, extremities_data.to_from_coor))
+
+        return unique_graph
+
+    @staticmethod
+    def find_node_id_by_coor(graph: MultiDiGraph, target_x: float, target_y: float):
+        """
+         finds the node in unique graph with the same coor
+        """
+        for node, data in graph.nodes(data=True):
+            if 'x' in data and 'y' in data and data['x'] == target_x and data['y'] == target_y:
+                return node
+        return None
+
+    @staticmethod
+    def snap_nodes_to_nodes(graph, threshold):
+        return consolidate_intersections(G=graph, rebuild_graph=True,
+                                         tolerance=threshold, dead_ends=False)
+
+    @staticmethod
+    def snap_nodes_to_edges(graph, threshold):
+        pass
+
+
+@dataclass
+class ExtremitiesData:
+    from_id: int = None
+    to_id: int = None
+    from_to_id: Union[None, tuple] = None
+    to_from_id: Union[None, tuple] = None
+    from_to_coor: Union[None, tuple] = None
+    to_from_coor: Union[None, tuple] = None
+
+    @staticmethod
+    def get_extremities_data_for_sub_graph(from_node_id: int, to_node_id: int, sub_graph: MultiDiGraph,
+                                           graph: MultiDiGraph, shared_elements: Union[set, None]):
+        """Both extremities should be in the unique_graph still makes an edge between similar node to u (the node
+        with u coordinates and different id, included in the unique_graph) and v Here, sub_graph is the unique_graph
+        and graph is complex_graph Shared elements are shared btw sub_graph and graph, which are elements to include
+        when dropping duplicates"""
+        if shared_elements is None:
```

Comment:
Remember what was mentioend earlier about the expected input. Don't initialize what the user has not given, raise errors instead!